insonmnia/hardware/hardware.go

```diff
@@ -146,56 +147,141 @@ func (m *benchValue) Value() uint64 {
+		if wasSet {
+			panic(fmt.Errorf("duplicate benchmark with id %d and type none", bench.ID))
 		}
-		to[id].Set(result)
 	case sonm.SplittingAlgorithm_PROPORTIONAL:
-		to[id].Add(uint64(float64(result) * proportion))
+		target.Result += uint64(float64(bench.Result) * proportion)
 	case sonm.SplittingAlgorithm_MAX:
-		if result >= to[id].Value() {
-			to[id].Set(result)
+		if bench.Result >= target.Result {
+			target.Result = bench.Result
 		}
 	case sonm.SplittingAlgorithm_MIN:
-		if !to[id].IsSet() {
-			to[id].Set(result)
-		} else {
-			if result < to[id].Value() {
-				to[id].Set(result)
+		if wasSet && bench.Result < target.Result {
+			target.Result = bench.Result
+		}
+	}
+}
+
+func (h *Hardware) ResourcesToBenchmarks(resources *sonm.AskPlanResources) (*sonm.Benchmarks, error) {
+	benchMap, err := h.ResourcesToBenchmarkMap(resources)
+	if err != nil {
+		return nil, err
+	}
+	var maxId uint64
+	for id := range benchMap {
+		if id > maxId {
+			maxId = id
+		}
+	}
+	resultBenchmarks := make([]uint64, maxId)
+	for k, v := range benchMap {
+		resultBenchmarks[k] = v.GetResult()
+	}
+	return sonm.NewBenchmarks(resultBenchmarks)
+}
+
+// This one is very similar to ResourcesToBenchmarkMap
+// except it stores corresponding benchmarks in corresponding devices.
+// This can not be simply reused due to MIN splitting algorithm (e.g. GPU mem - in this case it is stored separately)
+// TODO: find a way to refactor all this shit.
+func (h *Hardware) LimitTo(resources *sonm.AskPlanResources) (*Hardware, error) {
+	hardware := &Hardware{
+		CPU: &sonm.CPU{Device: h.CPU.Device, Benchmarks: map[uint64]*sonm.Benchmark{}},
+		GPU: []*sonm.GPU{},
+		RAM: &sonm.RAM{Device: h.RAM.Device, Benchmarks: map[uint64]*sonm.Benchmark{}},
+		Network: &sonm.Network{
+			In:       h.Network.In,
+			Out:      h.Network.Out,
+			Overlay:  h.Network.Overlay,
+			Incoming: h.Network.Incoming,
+			Outbound: h.Network.Outbound,
+		},
+		Storage: &sonm.Storage{Device: h.Storage.Device, Benchmarks: map[uint64]*sonm.Benchmark{}},
+	}
+
+	for _, gpu := range h.GPU {
+		hardware.GPU = append(hardware.GPU, &sonm.GPU{Device: gpu.Device, Benchmarks: map[uint64]*sonm.Benchmark{}})
+	}
+
+	for _, hash := range resources.GetGPU().GetHashes() {
+		found := false
+		for idx, gpu := range h.GPU {
+			if gpu.GetDevice().GetHash() == hash {
+				for _, bench := range gpu.GetBenchmarks() {
+					insertBench(hardware.GPU[idx].Benchmarks, bench, 1.0)
+				}
+				found = true
+				break
 			}
 		}
+		if !found {
+			return nil, fmt.Errorf("unknown hash in passed resources: %s", hash)
+		}
+	}
+
+	proportion := float64(resources.GetCPU().GetCorePercents()) / float64(h.CPU.GetDevice().GetCores()) / 100
+	for _, bench := range h.CPU.GetBenchmarks() {
+		insertBench(hardware.CPU.Benchmarks, bench, proportion)
+	}
 
+	proportion = float64(resources.GetStorage().GetSize().GetBytes()) / float64(h.Storage.GetDevice().GetBytesAvailable())
```

Comment:
seems like division by zero may happen

Reply:
it's checked for Nan and Inf, no worries